projects/user_publication/deduplicate.py

```diff
@@ -1,53 +1,87 @@
+          that is flagged as duplicate values are ID of the original publication
     """
-    # Exclude all publications with the status of 'rejected' or 'duplicate'
-    exclude_status = Q(status=Publication.STATUS_REJECTED)
-    exclude_status |= Q(status=Publication.STATUS_DUPLICATE)
+    # return a mapping of flagged duplicates and their original publications
+    duplicate_with_their_original_pubs_map = defaultdict(list)
 
     # Get a list of publications to check for duplicates
-    if pub_model:
-        pubs_to_check_duplicates = [pub_model]
+    if pubs:
+        pubs_to_check_duplicates = pubs
     else:
-        # get the publications in reverse so latest can be checked against
-        # old ones.
-        pubs_to_check_duplicates = Publication.objects.exclude(exclude_status).order_by("-id")
+        # get the publications in reverse so latest can be checked against old ones
+        # reviewed=True are the ones that are reviewed and are concrete non-duplicates
+        pubs_to_check_duplicates = Publication.objects.filter(reviewed=False).order_by("-id")
 
     # Loop through each publication to check for duplicates
     for pub1 in pubs_to_check_duplicates:
         # Get a subset of publications with id less than the publication at question
         # and of same year as the publication at question to check against
-        pubs_to_check_against = Publication.objects.filter(id__lt=pub1.id, year=pub1.year).order_by('-id')
-        pubs_to_check_against = pubs_to_check_against.exclude(exclude_status)
+        if not pub1.year:
+            logger.info(f"{pub1.title} does not have year - ignoring")
+            continue
+        # publications without ID means they are not in the database yet. So get all Publications
+        # that are older than pub1
+        if not pub1.id:
+            pubs_to_check_against = Publication.objects.filter(year=pub1.year).order_by('-id')
+        else:
+            pubs_to_check_against = Publication.objects.filter(id__lt=pub1.id, year=pub1.year).order_by('-id')
 
         # Loop through each subset publication to check against for duplicates
         for pub2 in pubs_to_check_against:
             # Check if pub1 and pub2 are similar enough to be flagged as duplicates
-            if PublicationUtils.is_pub_similar(pub1, pub2):
-                # Flag pub1 as a duplicate if it is not a dry run
-                if not dry_run:
-                    with transaction.atomic():
-                        pub1.status = Publication.STATUS_DUPLICATE
-                        pub1.save()
-                # Log the publications that have been flagged as duplicates
-                logger.info(f"{pub1.title} is flagged duplicate")
-                # break the inner loop and continue the outer loop to check duplicate for
-                # a new publication
-                break
+            if not PublicationUtils.is_pub_similar(pub1, pub2):
```

Comment:
I think it would be clearer to have `if PublicationUtils.is_pub_similar(pub1, pub2):` and then moving the status update to inside the if, rather than use continue. It takes more thought to understand the negative condition + continue.